contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,381 @@
+
+    // UNISWAP_ROUTER is Uniswap's periphery contract for conducting trades. Using this contract is gas inefficient and should only used for convenience i.e. admin functions
+    address private constant UNISWAP_ROUTER =
+        0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D;
+
+    address private constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+    address private constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+
+    constructor(
+        address _optionViews,
+        address _optionRegistry
+    ) {
+        require(_optionViews != address(0), "!_optionViews");
+        require(_optionRegistry != address(0), "!_optionRegistry");
+        optionViews = IOptionViews(_optionViews);
+        optionRegistry = IOptionRegistry(_optionRegistry);
+    }
+
+    receive() external payable {}
+
+    function protocolName() external pure override returns (string memory) {
+        return _name;
+    }
+
+    function nonFungible() external pure override returns (bool) {
+        return _nonFungible;
+    }
+
+    function purchaseMethod()
+        external
+        pure
+        override
+        returns (ProtocolAdapterTypes.PurchaseMethod)
+    {
+        return ProtocolAdapterTypes.PurchaseMethod.Contract;
+    }
+
+    /**
+     * @notice Check if an options contract exist based on the passed parameters.
+     * @param optionTerms is the terms of the option contract
+     */
+    function optionsExist(ProtocolAdapterTypes.OptionTerms calldata optionTerms)
+        external
+        view
+        override
+        returns (bool)
+    {
+        return address(lookupCToken(optionTerms)) != address(0);
+    }
+
+    /**
+     * @notice Get the options contract's address based on the passed parameters
+     * @param optionTerms is the terms of the option contract
+     */
+    function getOptionsAddress(
+        ProtocolAdapterTypes.OptionTerms calldata optionTerms
+    ) external view override returns (address) {
+        return address(lookupCToken(optionTerms));
+    }
+
+    /**
+     * @notice Gets the premium to buy `purchaseAmount` of the option contract in ETH terms.
+     */
+    function premium(
+      ProtocolAdapterTypes.OptionTerms calldata optionTerms,
+      uint256 purchaseAmount
+    )
+        external
+        view
+        override
+        returns (uint256 cost)
+    {
+        address tokenAddress = lookupCToken(optionTerms);
+
+        require(tokenAddress != address(0), "Must be valid option terms!");
+
+        //get token
+        IOptionToken token = IOptionToken(tokenAddress);
+
+        // Get strike index, whether is long
+        IOptionRegistry.OptionDetails memory optionType = optionRegistry.getOptionDetails(token);
+
+        //get market
+        IOptionMarket market = IOptionMarket(token.market());
+
+        cost = optionViews.getBuyOptionCost(market, optionType.isLongToken, optionType.strikeIndex, purchaseAmount.mul(10 ** token.decimals()).div(10 ** 18));
```

Comment:
We can avoid calling `token.decimals` here if we know beforehand how many decimals they are. Able to hardcode the `decimals` in the contract itself

Reply:
so you recommend a mapping from address -> uint256 decimals ?

problem is each token has its own decimal. for base token usdc an option token will be 6. wbtc and eth is 8 and 18 respectively so it will depend on the current token interacting with.

Reply:
I see. It's okay then, disregard this comment.